src/append.jl

```diff
@@ -0,0 +1,214 @@
+        elseif compress === :zstd
+            compress = ZSTD_COMPRESSOR
+        elseif compress isa Symbol
+            throw(ArgumentError("unsupported compress keyword argument value: $compress. Valid values include `:lz4` or `:zstd`"))
+        end
+        seek(io, length(bytes) - 8) # overwrite last 8 bytes of last empty message footer
+        append(io, tbl, arrow_schema, compress, largelists, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    end
+
+    return file
+end
+
+function append(io::IO, source, arrow_schema, compress, largelists, denseunions, dictencode, dictencodenested, alignment, maxdepth, ntasks)
+    sch = Ref{Tables.Schema}(arrow_schema)
+    msgs = OrderedChannel{Message}(ntasks)
+    dictencodings = Dict{Int64, Any}() # Lockable{DictEncoding}
+    # build messages
+    blocks = (Block[], Block[])
+    # start message writing from channel
+    threaded = ntasks > 1
+    tsk = threaded ? (Threads.@spawn for msg in msgs
+        Base.write(io, msg, blocks, sch, alignment)
+    end) : (@async for msg in msgs
+        Base.write(io, msg, blocks, sch, alignment)
+    end)
+    anyerror = Threads.Atomic{Bool}(false)
+    errorref = Ref{Any}()
+    @sync for (i, tbl) in enumerate(Tables.partitions(source))
+        if anyerror[]
+            @error "error writing arrow data on partition = $(errorref[][3])" exception=(errorref[][1], errorref[][2])
+            error("fatal error writing arrow data")
+        end
+        @debug 1 "processing table partition i = $i"
+        tbl_schema = Tables.schema(tbl)
+
+        if !is_equivalent_schema(arrow_schema, tbl_schema)
+            throw(ArgumentError("Table schema does not match existing arrow file schema"))
+        end
+
+        if threaded
+            Threads.@spawn process_partition(tbl, dictencodings, largelists, compress, denseunions, dictencode, dictencodenested, maxdepth, msgs, alignment, i, sch, errorref, anyerror)
+        else
+            @async process_partition(tbl, dictencodings, largelists, compress, denseunions, dictencode, dictencodenested, maxdepth, msgs, alignment, i, sch, errorref, anyerror)
+        end
+    end
+    if anyerror[]
+        @error "error writing arrow data on partition = $(errorref[][3])" exception=(errorref[][1], errorref[][2])
+        error("fatal error writing arrow data")
+    end
+    # close our message-writing channel, no further put!-ing is allowed
+    close(msgs)
+    # now wait for our message-writing task to finish writing
+    wait(tsk)
+
+    Base.write(io, Message(UInt8[], nothing, 0, true, false, Meta.Schema), blocks, sch, alignment)
+
+    return io
+end
+
+function table_info(file::String; kwargs...)
+    open(file) do io
+        table_info(io; file=true, kwargs...)
+    end
+end
+table_info(io::IO; file::Bool=false, kwargs...) = table_info(file ? Mmap.mmap(io) : Base.read(io); kwargs...)
+function table_info(bytes::Vector{UInt8}; convert::Bool=true)
```

Comment:
Hmmmm, I don't like duplicating all this code from `Arrow.Stream`/`Arrow.Table`. I think we should modify those to store whatever is needed: schema, compression, etc. I'd much rather add a field or two to `Arrow.Stream` rather than have all this duplicate code.

Reply:
Yes, good idea. Will make the change.

Reply:
done in https://github.com/JuliaData/Arrow.jl/pull/160/commits/24d5920cacd845c13d18e3dbf640ba4a64dd552c